core/lib/dal/src/consensus_dal.rs

```diff
@@ -480,6 +461,47 @@ impl ConsensusDal<'_, '_> {
             .number
             .map(|number| attester::BatchNumber(number as u64)))
     }
+
+    /// Get the numbers of L1 batches which do not have a corresponding L1 quorum certificate
+    /// and need signatures to be gossiped and collected.
+    ///
+    /// On the main node this means every L1 batch, because we need QC over all of them to be
+    /// able to submit them to L1. Replicas don't necessarily have to have the QC, because once
+    /// the batch is on L1 and it's final, they can get the batch from there and don't need the
+    /// attestations.
+    pub async fn unsigned_batch_numbers(
+        &mut self,
+        min_batch_number: attester::BatchNumber,
+    ) -> DalResult<Vec<attester::BatchNumber>> {
+        Ok(sqlx::query!(
+            r#"
+            SELECT
+                b.number
+            FROM
+                l1_batches b
+            WHERE
+                b.number >= $1
+                AND NOT EXISTS (
+                    SELECT
+                        1
+                    FROM
+                        l1_batches_consensus c
+                    WHERE
+                        c.l1_batch_number = b.number
+                )
+            ORDER BY
+                b.number
```

Comment:
I'd probably do a join and then filter out values for which there is a matching row in consensus table (`WHERE l1_batches_consensus. l1_batch_number IS NULL`) - it's arguably more idiomatic

Reply:
From my DB days, these two in an ideal test plan both turned into ANTI-JOIN:
* `select * from a where not exists (select 1 from b where b.a_id = a.id)`
* `select a.* from a left outer join b on a.id = b.a_id where b.id is null`

I don't have a preference, if but just to emphasize it has to be a _left outer join_ otherwise it would do the opposite. If you understand it the same way I'm happy to rewrite, no problem.

Reply:
I'd expect the `not exists` syntax to have some advantages if one really wants to achieve an _anti join_ in that:
* you can't accidentally use the columns of the right side in the select
* under no circumstances will it duplicate data from the left side
* the intent is very clear (to my eyes at least)

I'd assume the `left outer join` syntax might require a foreign key to realise it's actually an _anti join_. 

In Postgres they should be equivalent in terms of plan, if everything goes the way it should. I think I started using this syntax because it was more likely to result in the right plan in Oracle, but not sure why that was. Maybe just some general advice I saw somewhere.

Reply:
Refactored to use left join in https://github.com/matter-labs/zksync-era/pull/2410